reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt

```diff
@@ -0,0 +1,626 @@
+    }
+
+    @Test
+    fun mustUseAtomicOperations_when_updateFeatureContext_addNewValues() {
+        // Given
+        val fakeKeyValues1 = forge.aMap(
+            size = forge.anInt(
+                min = 1,
+                max = 10
+            )
+        ) { forge.anAlphabeticalString() to forge.anAlphabeticalString() }
+        val fakeKeyValues2 = forge.aMap(
+            size = forge.anInt(
+                min = 1,
+                max = 10
+            )
+        ) { forge.anAlphabeticalString() to forge.anAlphabeticalString() }
+        val expectedKeyValues = fakeKeyValues1 + fakeKeyValues2
+        val countDownLatch = CountDownLatch(2)
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                fakeKeyValues1.forEach { (key, value) ->
+                    it[key] = value
+                }
+            }
+        }.start()
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                fakeKeyValues2.forEach { (key, value) ->
+                    it[key] = value
+                }
+            }
+        }.start()
+        testedInternalSdkCore?.setContextUpdateReceiver(fakeFeatureName) { _, _ -> countDownLatch.countDown() }
+        countDownLatch.await(SHORT_WAIT_MS, TimeUnit.MILLISECONDS)
+
+        // When
+        val context = testedInternalSdkCore?.getDatadogContext()
+
+        // Then
+        assertThat(context?.featuresContext?.entries).hasSize(1)
+        val featureContext = context?.featuresContext?.get(fakeFeatureName)
+        assertThat(featureContext).containsExactlyInAnyOrderEntriesOf(expectedKeyValues)
+    }
+
+    @Test
+    fun mustUseAtomicOperations_when_updateFeatureContext_modifyValues() {
+        // Given
+        val fakeKeyValues1 = forge.aMap(
+            size = forge.anInt(
+                min = 1,
+                max = 10
+            )
+        ) { forge.anAlphabeticalString() to forge.anAlphabeticalString() }
+        val fakeKeyValues2 = forge.aMap(
+            size = forge.anInt(
+                min = 1,
+                max = 10
+            )
+        ) { forge.anAlphabeticalString() to forge.anAlphabeticalString() }
+        val fakeModifiedValues = fakeKeyValues2.mapValues { (_, value) -> forge.anAlphabeticalString() }
+        val expectedKeyValues = fakeKeyValues1 + fakeModifiedValues
+        val countDownLatch = CountDownLatch(3)
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                fakeKeyValues1.forEach { (key, value) ->
+                    it[key] = value
+                }
+            }
+        }.start()
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                fakeKeyValues2.forEach { (key, value) ->
+                    it[key] = value
+                }
+            }
+        }.start()
+        // Give some time to the first 2 threads to start
+        Thread.sleep(100)
```

Comment:
minor: to be more deterministic it is better to use another latch, but I guess 100ms should be enough.